check/src/completion.rs

```diff
@@ -1,78 +1,81 @@
 //! Primitive auto completion and type quering on ASTs
-use std::iter::once;
 
-use base::ast;
-use base::scoped_map::ScopedMap;
+use std::iter::once;
 use std::cmp::Ordering;
 
-use base::ast::{DisplayEnv, Location, Typed};
-use base::symbol::Symbol;
-use base::types::{Type, TcType};
+use base::ast::{self, DisplayEnv, Expr, LExpr, Location, LPattern, Pattern, TcIdent, Typed};
 use base::instantiate;
+use base::scoped_map::ScopedMap;
+use base::symbol::Symbol;
+use base::types::{TcType, Type, TypeEnv};
 
 trait OnFound {
-    fn on_ident(&mut self, ident: &ast::TcIdent<Symbol>) {
+    fn on_ident(&mut self, ident: &TcIdent<Symbol>) {
         let _ = ident;
     }
-    fn on_expr(&mut self, expr: &ast::LExpr<ast::TcIdent<Symbol>>) {
-        let _ = expr;
-    }
-    fn on_pattern(&mut self, pattern: &ast::LPattern<ast::TcIdent<Symbol>>) {
+
+    fn on_pattern(&mut self, pattern: &LPattern<TcIdent<Symbol>>) {
         let _ = pattern;
     }
 
-    fn expr(&mut self, expr: &ast::LExpr<ast::TcIdent<Symbol>>);
-    fn ident(&mut self, context: &ast::LExpr<ast::TcIdent<Symbol>>, ident: &ast::TcIdent<Symbol>);
+    fn expr(&mut self, expr: &LExpr<TcIdent<Symbol>>);
+
+    fn ident(&mut self, context: &LExpr<TcIdent<Symbol>>, ident: &TcIdent<Symbol>);
 }
 
-struct GetType(Option<TcType>);
-impl OnFound for GetType {
-    fn expr(&mut self, expr: &ast::LExpr<ast::TcIdent<Symbol>>) {
-        self.0 = Some(expr.type_of());
+struct GetType<'a, T: TypeEnv + 'a> {
+    env: &'a T,
```

Comment:
You could have `env: T` instead and remove the lifetime parameter if the implementation for `&T: TypeEnv` are kept. Then `GetType<&TypeEnv>` would work for instance.